### ISSUE TITLE
Support Routing All Kubernetes Master and Worker Nodes Through AWS NAT Gateway

### ISSUE THREAD

**Issue by sostheim** [samsung-cnct/k2#634]
As a Cluster Operator on AWS Cloud Provider infrastructure, I want to be able to route all of my node traffic to and from the infrastructure via a an [AWS NAT Gateway](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/vpc-nat-gateway.html).

**Comment by sostheim** [comment]
This may be as simple as removing all external IP address from both master and worker nodes, as described in #575 

However, this may not be the only element to facilitating the the NAT traversal.  This ticket is for the investigation necessary to support this capability.

**Comment by coffeepac** [comment]
I think that all traffic goes through the NAT/ELBs now.  The public ips are, I think, only for allowing ssh access. 

Easy test:  using k2 modify the nodePools templates to have `false` for external ips and see what happens.

**Comment by alejandroEsc** [comment]
I have been doing that, but running into issues getting kubernetes started. going to try to redeploy in a bit, leaving special nodes ssh-able.

**Comment by coffeepac** [comment]
okay.  if you don't have any luck I'll pick up the other ticket and sort out what's going wrong.

**Comment by alejandroEsc** [comment]
I think the issue may be that because the node is private it has not access to the internet gateway, I could be wrong of course: Here is logs of what i think is the start of the failure:

```
Jul 17 20:06:41 ip-10-0-127-41.us-west-2.compute.internal systemd[1]: Failed to start Setup Network Environment.
Jul 17 20:06:41 ip-10-0-127-41.us-west-2.compute.internal systemd[1]: kraken-networking.service: Unit entered failed state.
Jul 17 20:06:41 ip-10-0-127-41.us-west-2.compute.internal systemd[1]: kraken-networking.service: Failed with result 'exit-code'.
Jul 17 20:06:41 ip-10-0-127-41.us-west-2.compute.internal systemd[1]: Starting Setup Network Environment...
Jul 17 20:07:41 ip-10-0-127-41.us-west-2.compute.internal rkt[893]: run: discovery failed```

Ill go ahead and post my PR in a bit so you can take a look of what I have tried. And yes being public is the default AND i had to make most of these configuration changes optional or i think we break backwards compatibility.

**Comment by alejandroEsc** [comment]
I think I am way over thinking this but I have concerns about guiding users to a public/private topology and how much handholding we expect. My suspicion is that they would need handholding period, but here is a suggestion about how this would be done:

- A copule of subnets per AZ, one public one private
- public subnets reserved for aws master nodes
- private subnets for workers and the like (etcd etc..)
- to allow pulls for private nodes, they need an internet gateway,which could be NAT-ted from the public subnets. 

Is this  kinda/sorta what we are looking for?

This leads into questions about what is the network for comcast like and if they have an internetgarteway NAT-ted to their nodes. My suspicions is that they do. 

Now that is a bit aside from what comcast is asking, which is to just set nodes to public/private. I am partially there (items checked off)r:

Cases:
- if Subnet is private
  - [X] each node type (master, node, special, etcd) can be public if enablePublicIps is set on their particular node provider (open to suggestion if you want to move this) 
  - [X] each node type (master, node, special, etcd) can be private if enablePublicIps is set on their particular node provider (open to suggestion if you want to move this) 

- if Subnet is public
  - [X] each node type (master, node, special, etcd) can be public if enablePublicIps is set on their particular node provider (open to suggestion if you want to move this) 
  - [ ] each node type (master, node, special, etcd) can be private if enablePublicIps is set on their particular node provider (open to suggestion if you want to move this)

**Comment by alejandroEsc** [comment]
FYI, testing the last item and seeing how k2 performs.

**Comment by alejandroEsc** [comment]
another bug on aws-provider affects etcd nodes: https://github.com/samsung-cnct/cts-k8s-2017/issues/39

**Comment by alejandroEsc** [comment]
Private networking works but so far only in the case where a vpc has been set up by hand, the only thing we are contingent on is the terraform-provider-aws setting things to private to work, but because we would presumably launch to a private subnet we dont really need this case at the moment and we can continue.

Couple of facts to keep in mind:
- So far this would only work on the bring your own vpc case, because you need to set up public and private subnets that can further be consumed as launching places for the 
- The routing and networking works as long as we have two subnets in each az: one public, one private; then manually configure the nodes that will be in the private subnet (essentially all of k8s). 
- ELB for master nodes, are by default, placed in the same subnet of the master nodes. We need to be able to remove this default behavior and override it to launch at the public subnet. route 53 entries should to be adjusted accordingly.
- Less importantly but we have to keep in mind that we need to turn on for the vpc the hostname resolution, the expected resolver needs to be in the cidr+2 address, not a big deal since we default this in our own vpc case, but in the BYOVPC that needs to be clear somewhere and if not set as optional. 

Suggestion:

We close this ticket. 

Open new tickets for a bastion and a config level ELB entry. Update documentation as needed. describing what a k2 private network topology looks like at present.

**Comment by coffeepac** [comment]
@alejandroEsc that seems like a reasonable path forward.  please open the follow-on issues, reference them back to here and close this issue when complete.

**Comment by leahnp** [comment]
@sostheim @coffeepac @alejandroEsc I am confused what the action item on this ticket is - seems research was done and we decided to break the work up into other tickets?

**Comment by alejandroEsc** [comment]
@leahnp correct. The prelim research is complete, changes committed a while ago that allow support for the comcast case, next items to do:

These are my suggestion next items/course of action
- [ ]  Complete work on Bastion : https://github.com/samsung-cnct/k2/issues/659
- [ ] Open ticket for a bastion with an ELB
- [ ] Open ticket for creating NAT-gateways and appropriate route tables.
- [ ] Open ticket on Creating k2 documentation discussing  private topologies and options.
- [ ] Open ticket on updating new config for private topologies showing available options
- [ ] Update k2 and k2cli so that --generate-private(-topology) will give you the aforementioned config file. 

Because they are really related to a private topology story, i would rather see a new story ticket created with the list above (feel free to change) and work off that ticket. I do feel this story is complete.

**Comment by leahnp** [comment]
Ok - @sostheim do you want to close this ticket and open new issues for this approach?

**Comment by leahnp** [comment]
@coffeepac did we want to close this ticket?

**Comment by coffeepac** [comment]
looks done.  outoutout

**Comment by sostheim** [comment]
@leahnp - sorry, missed your first ask about closing this ticket 9 day's ago.  Thanks for resolving.
